moa/src/main/java/moa/streams/generators/SineGenerator.java

```diff
@@ -0,0 +1,204 @@
+/*
+ *    SineGenerator.java
+ *    Copyright (C) 2016 Instituto Federal de Pernambuco
+ *    @author Paulo Gonçalves (paulogoncalves@recife.ifpe.edu.br)
+ *
+ *    This program is free software; you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation; either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *    
+ */
+package moa.streams.generators;
+
+import com.github.javacliparser.FlagOption;
+import com.github.javacliparser.IntOption;
+import com.yahoo.labs.samoa.instances.Attribute;
+import com.yahoo.labs.samoa.instances.DenseInstance;
+import com.yahoo.labs.samoa.instances.Instance;
+import com.yahoo.labs.samoa.instances.Instances;
+import com.yahoo.labs.samoa.instances.InstancesHeader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import moa.core.InstanceExample;
+
+import moa.core.ObjectRepository;
+import moa.options.AbstractOptionHandler;
+import moa.streams.InstanceStream;
+import moa.tasks.TaskMonitor;
+
+/**
+ * Stream generator for the problem of identifying a Sine curve. Based on
+ * Sine1, Sine2, Sinirrel1, and Sinirrel2 as proposed by
+ * "Gama, Joao, et al. "Learning with drift detection." Advances in artificial 
+ * intelligence–SBIA 2004. Springer Berlin Heidelberg, 2004. 286-295."
+ *
```

Comment:
We may also add the original explanation to make the code easier to understand:

SINE1. Abrupt concept drift, noise-free examples. It has
two relevant attributes. Each attributes has values uniformly distributed in
[0; 1]. In the first context all points below the curve y = sin(x) are classified
as positive. After the context change the classification is reversed.
2. SINE2. The same two relevant attributes. The classification function is y <
0,5 + 0,3 sin(3 PI x). After the context change the classification is reversed.
3. SINIRREL1. Presence of irrelevant attributes. The same classification
function of SINE1 but the examples have two more random attributes
with no influence on the classification function.
4. SINIRREL2. The same classification function of SINE2 but the examples
have two more random attributes with no influence on the classification
function.